File: Framework/ClaseBaseLN/BaseGenericLN.vb
#Region "importaciones"

Imports Framework.LogicaNegocios.Transacciones
Imports Framework.DatosNegocio
Imports Framework.AccesoDatos.MotorAD.LN
Imports Framework.AccesoDatos.MotorAD.AD

#End Region

Public Class BaseGenericLN
    Inherits BaseTransaccionLN


#Region "Contructor"
    Sub New(ByVal pTL As ITransaccionLogicaLN, ByVal pRec As IRecursoLN)
        MyBase.New(pTL, pRec)
    End Sub
#End Region

    ''' <summary>
    ''' Solo se puede utilizar si el código es generico sin lógica
    ''' <param name="id"></param>
    ''' </summary>
    Protected Function Recuperar(Of T As EntidadDN)(ByVal huella As Framework.DatosNegocio.IHuellaEntidadDN) As T


        If Not huella.TipoEntidadReferida Is GetType(T) Then
            Throw New LogicaNegocios.ApplicationExceptionLN("los tipos no son compatibles")
        End If

        Return Recuperar(Of T)(huella.IdEntidadReferida)

    End Function

    Protected Function Recuperar(Of T As EntidadDN)(ByVal id As String) As T
        Return Recuperar(Of T)(id, mTL)
    End Function

    Private Function Recuperar(Of T As EntidadDN)(ByVal id As String, ByVal trans As ITransaccionLogicaLN) As T
        Dim tlproc As ITransaccionLogicaLN = Nothing
        Dim ctd As CTDLN
        Dim gestor As GestorInstanciacionLN
        Dim result As T = Nothing

        ctd = New CTDLN()

        Try
            ctd.IniciarTransaccion(trans, tlproc)

            gestor = New GestorInstanciacionLN(tlproc, Me.mRec)
            result = gestor.Recuperar(Of T)(id)

            tlproc.Confirmar()
            Return result
        Catch e As Exception
            If (Not (tlproc Is Nothing)) Then
                tlproc.Cancelar()
            End If
            Throw
        End Try

    End Function

    Protected Function RecuperarLista(Of T As EntidadBaseDN)() As List(Of T)
        Dim tlproc As ITransaccionLogicaLN = Nothing
        Dim aMAD As AccesorMotorAD = Nothing

        Try
            tlproc = ObtenerTransaccionDeProceso()
            aMAD = New AccesorMotorAD(tlproc, mRec, New ConstructorAL(GetType(T)))

            '  Dim alIDs As ArrayList = aMAD.BuscarGenericoIDS("tl" + GetType(T).Name, Nothing)
            Dim alIDs As ArrayList = aMAD.BuscarGenericoIDS(GetType(T), Nothing)

            Dim objetos As List(Of T) = Me.RecuperarLista(Of T)(alIDs)

            tlproc.Confirmar()

            Return objetos

        Catch ex As Exception
            If (Not (tlproc Is Nothing)) Then
                tlproc.Cancelar()
            End If
            Throw
        End Try

    End Function

    Protected Function RecuperarLista(Of T As EntidadBaseDN)(ByVal listaIDs As ArrayList) As List(Of T)
        Dim tlproc As ITransaccionLogicaLN = Nothing
        Dim gestor As GestorInstanciacionLN

        Try
            tlproc = ObtenerTransaccionDeProceso()

            gestor = New GestorInstanciacionLN(tlproc, Me.mRec)

            Dim alObj As ArrayList = CType(gestor.Recuperar(listaIDs, GetType(T), Nothing), ArrayList)
            Dim objetos As List(Of T) = New List(Of T)(alObj.Count)
            For i As Integer = 0 To alObj.Count - 1
                objetos.Add(CType(alObj(i), T))
            Next

            tlproc.Confirmar()

            Return objetos

        Catch ex As Exception
            If (Not (tlproc Is Nothing)) Then
                tlproc.Cancelar()
            End If
            Throw

        End Try

    End Function

    Protected Function RecuperarListaCondicional(Of T As EntidadDN)(ByVal constructor As IConstructorBusquedaAD) As List(Of T)
        Dim amd As AccesorMotorAD = Nothing
        Dim tlproc As ITransaccionLogicaLN = Nothing
        Dim ctd As CTDLN
        Dim alIds As ArrayList = Nothing
        Dim gestor As GestorInstanciacionLN

        ctd = New CTDLN()
        Try
            ctd.IniciarTransaccion(mTL, tlproc)

            amd = New AccesorMotorAD(tlproc, mRec, constructor)
            alIds = amd.BuscarGenericoIDS(GetType(T))
            gestor = New GestorInstanciacionLN(tlproc, Me.mRec)
            Dim alObj As ArrayList = CType(gestor.Recuperar(alIds, GetType(T), Nothing), ArrayList)
            Dim objetos As List(Of T) = New List(Of T)(alObj.Count)
            For i As Integer = 0 To (alObj.Count - 1)
                objetos.Add(CType(alObj(i), T))
            Next

            tlproc.Confirmar()
            Return objetos
        Catch ex As Exception
            If (Not (tlproc Is Nothing)) Then
                tlproc.Cancelar()
            End If
            Throw
        End Try

    End Function

    ''' <summary>
    ''' Solo se puede utilizar si el código es genérico sin lógica
    ''' </summary>
    ''' <typeparam name="T"></typeparam>
    ''' <param name="dato"></param>
    Protected Function Guardar(Of T As EntidadDN)(ByVal dato As T) As T
        Dim tlproc As ITransaccionLogicaLN = Nothing
        Dim ctd As CTDLN
        Dim gestor As GestorInstanciacionLN

        ctd = New CTDLN()
        Try
            ctd.IniciarTransaccion(mTL, tlproc)

            gestor = New GestorInstanciacionLN(tlproc, Me.mRec)
            gestor.Guardar(dato)

            tlproc.Confirmar()

            Return dato
        Catch ex As Exception
            If (Not (tlproc Is Nothing)) Then
                tlproc.Cancelar()
            End If
            Throw
        End Try

    End Function

    ''' <summary>
    ''' Solo se puede utilizar si el código es genérico sin lógica
    ''' </summary>
    ''' <param name="lista">Lista de objetos a guardar</param>
    Protected Sub GuardarLista(Of T As EntidadDN)(ByVal lista As List(Of T))
        Dim tlproc As ITransaccionLogicaLN = Nothing
        Dim ctd As CTDLN
        Dim gestor As GestorInstanciacionLN

        ctd = New CTDLN()
        Try
            ctd.IniciarTransaccion(mTL, tlproc)

            gestor = New GestorInstanciacionLN(tlproc, Me.mRec)
            For i As Integer = 0 To lista.Count - 1
                gestor.Guardar(lista(i))
            Next

            tlproc.Confirmar()
        Catch ex As Exception
            If (Not (tlproc Is Nothing)) Then
                tlproc.Cancelar()
            End If
            Throw
        End Try

    End Sub

    ''' <summary>
    ''' Solo se puede utilizar si el código es genérico sin lógica
    ''' </summary>
    ''' <typeparam name="T"></typeparam>
    ''' <param name="id"></param>
    Protected Sub Baja(Of T As EntidadDN)(ByVal id As String)
        Dim dato As IDatoPersistenteDN = CType(Recuperar(Of T)(id), IDatoPersistenteDN)
        Dim tlproc As ITransaccionLogicaLN = Nothing
        Dim ctd As CTDLN
        Dim gestor As GestorInstanciacionLN

        ctd = New CTDLN()
        Try
            ctd.IniciarTransaccion(mTL, tlproc)

            gestor = New GestorInstanciacionLN(tlproc, Me.mRec)
            dato.Baja = True
            gestor.Baja(dato)

            tlproc.Confirmar()
        Catch ex As Exception
            If (Not (tlproc Is Nothing)) Then
                tlproc.Cancelar()
            End If
            Throw
        End Try

    End Sub

    ''' <summary>
    ''' Solo se puede utilizar si el código es genérico sin lógica
    ''' <param name="id"></param>
    ''' </summary>
    Protected Sub Reactivar(Of T As EntidadDN)(ByVal id As String)
        Dim dato As IDatoPersistenteDN = CType(Recuperar(Of T)(id), IDatoPersistenteDN)
        dato.Baja = False
        Guardar(Of T)(CType(dato, T))
    End Sub

End Class

File: Framework/Transacciones/BaseGenericLN.cs
#region Importaciones

using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using Framework.AccesoDatos;
using Framework.AccesoDatos.MotorAD.AD;
using Framework.AccesoDatos.MotorAD.LN;
using Framework.DatosNegocio;

#endregion

namespace Framework.LogicaNegocios.Transacciones
{
    public abstract class BaseGenericLN
    {
        private GestorInstanciacionLN NuevoGestor()
        {
            return new GestorInstanciacionLN(Transaccion.Actual, Recurso.Actual);
        }

        private Ejecutor NuevoEjecutor()
        {
            return new Ejecutor(Transaccion.Actual, Recurso.Actual);
        }

        private AccesorMotorAD NuevoAccesorMotorAD(IConstructorBusquedaAD constructor)
        {
            return new AccesorMotorAD(Transaccion.Actual, Recurso.Actual, constructor);
        }

        #region Metodos

        protected void GenerarTablas<T>()
        {
            using (Transaccion tr = new Transaccion())
            {
                NuevoGestor().GenerarTablas2(typeof (T), null);

                tr.Confirmar();
            }
        }

        /// <summary>
        /// Solo se puede utilziar si el código es genérico sin lógica
        /// </summary>
        /// <typeparam name="T"></typeparam>
        /// <param name="dato"></param>
        protected void Guardar<T>(T dato) where T : EntidadDN
        {
            using (Transaccion tr = new Transaccion())
            {
                NuevoGestor().Guardar(dato);

                tr.Confirmar();
            }
        }

        /// <summary>
        /// Solo se puede utilziar si el código es genérico sin lógica
        /// </summary>
        /// <typeparam name="List<T>"></typeparam>
        /// <param name="lista"></param>
        protected void GuardarLista<T>(List<T> lista) where T : EntidadDN
        {
            using (Transaccion tr = new Transaccion())
            {
                GestorInstanciacionLN gestor = NuevoGestor();

                for (int i = 0; i < lista.Count; i++)
                    gestor.Guardar(lista[i]);

                tr.Confirmar();
            }
        }

        /// <summary>
        /// Solo se puede utilizar si el código es genérico sin lógica
        /// </summary>
        /// <typeparam name="T"></typeparam>
        /// <param name="id"></param>
        /// <returns></returns>
        protected T Recuperar<T>(string id) where T : EntidadDN
        {
            using (Transaccion tr = new Transaccion())
            {
                T result = NuevoGestor().Recuperar<T>(id);

                tr.Confirmar();

                return result;
            }
        }

        protected void Eliminar<T>(string id)
        {
            using (Transaccion tr = new Transaccion())
            {
                string nombreTipo = typeof (T).Name;

                Ejecutor ejec = NuevoEjecutor();
                string misql = "DELETE FROM tl" + nombreTipo + " WHERE ID = " + id;

                if (ejec.EjecutarNoConsulta(misql) == 0)
                    throw new NingunaFilaAfectadaException("No se ejecutó correctamente la eliminación del " +
                                                           nombreTipo + ". Ninguna fila afectada");

                tr.Confirmar();
            }
        }

        protected int EliminarRelacion(string table, string campo, string idCampo)
        {
            using (Transaccion tr = new Transaccion())
            {
                Ejecutor ejec = NuevoEjecutor();
                string misql = "DELETE FROM " + table + " WHERE " + campo + " = " + idCampo;

                int val = ejec.EjecutarNoConsulta(misql);

                tr.Confirmar();

                return val;
            }
        }


        /// <summary>
        /// Solo se puede utilziar si el código es genérico sin lógica
        /// </summary>
        /// <typeparam name="T"></typeparam>
        /// <param name="id"></param>
        protected void Reactivar<T>(string id) where T : EntidadDN
        {
            IDatoPersistenteDN dato = (IDatoPersistenteDN) Recuperar<T>(id);
            dato.Baja = false;
            Guardar<T>((T) dato);
        }

        /// <summary>
        /// Solo se puede utilizar si el código es genérico sin lógica
        /// </summary>
        /// <typeparam name="T"></typeparam>
        /// <param name="id"></param>
        protected void Baja<T>(string id) where T : EntidadDN
        {
            IDatoPersistenteDN dato = (IDatoPersistenteDN) Recuperar<T>(id);

            using (Transaccion tr = new Transaccion())
            {
                dato.Baja = true;
                NuevoGestor().Baja(dato);

                tr.Confirmar();
            }
        }

        protected List<T> RecuperarLista<T>() where T : EntidadBaseDN
        {
            using (Transaccion tr = new Transaccion())
            {
                AccesorMotorAD aMAD = NuevoAccesorMotorAD(new ConstructorAL(typeof(T)));

                //ArrayList alIDs = aMAD.BuscarGenericoIDS("tl" + typeof (T).Name, null);
                ArrayList alIDs = aMAD.BuscarGenericoIDS( typeof(T), null);

                ArrayList alObj = (ArrayList) NuevoGestor().Recuperar(alIDs, typeof (T), null);

                List<T> objetos = new List<T>(alObj.Count);
                for (int i = 0; i < alObj.Count; i++)
                    objetos.Add((T) alObj[i]);

                tr.Confirmar();

                return objetos;
            }
        }

        protected List<T> RecuperarListaCondicional<T>(IConstructorBusquedaAD constructor) where T : EntidadDN
        {
            using (Transaccion tr = new Transaccion())
            {
                AccesorMotorAD amd = NuevoAccesorMotorAD(constructor);

                ArrayList alIDs = amd.BuscarGenericoIDS(typeof(T));

                ArrayList alObj = (ArrayList) NuevoGestor().Recuperar(alIDs, typeof (T), null);

                List<T> objetos = new List<T>(alObj.Count);
                for (int i = 0; i < alObj.Count; i++)
                    objetos.Add((T) alObj[i]);

                tr.Confirmar();

                return objetos;
            }
        }

        protected List<T> RecuperarListaCondicionalConEntidadesRecuperadas<T>(IConstructorBusquedaAD constructor,
                                                                              IList recuperadas) where T : EntidadDN
        {
            using (Transaccion tr = new Transaccion())
            {
                AccesorMotorAD amd = NuevoAccesorMotorAD(constructor);

                ArrayList alIDs = amd.BuscarGenericoIDS(typeof(T));

                GestorInstanciacionLN gestor = NuevoGestor();

                //hace que los enlaces se hagan bien
                gestor.AñadirAColIntanciasRecuperadas(recuperadas);

                ArrayList alObj = (ArrayList) gestor.Recuperar(alIDs, typeof (T), null);

                List<T> objetos = new List<T>(alObj.Count);
                for (int i = 0; i < alObj.Count; i++)
                    objetos.Add((T) alObj[i]);

                tr.Confirmar();

                return objetos;
            }
        }

        protected delegate T ConstructorDN<T>(DataRow dr);

        protected List<T> RecuperarListaCondicionalConstructorManual<T>(IConstructorBusquedaAD constSQL,
                                                                        ConstructorDN<T> constDN)
        {
            using (Transaccion tr = new Transaccion())
            {
                AccesorMotorAD amd = NuevoAccesorMotorAD(constSQL);

                DataSet ds = null;
                ds = amd.BuscarGenerico(ref ds);

                if (ds == null || ds.Tables.Count != 1)
                    throw new ApplicationException("Error: No se ha podido hacer la consulta");

                List<T> lista = new List<T>();

                foreach (DataRow dr in ds.Tables[0].Rows)
                {
                    T dn = constDN(dr);
                    lista.Add(dn);
                }

                tr.Confirmar();

                return lista;
            }
        }

        #endregion
    }
}